### DISCUSSION TITLE
Chunking mechanic akin to Spring Batch chunking ?

### DISCUSSION THREAD

**Discussion by Celludriel** [jobrunr/jobrunr#856] [Q&A] [answered]
Hey,

We are investigating if Jobrunr is a good fit for our company so we can improve on scheduling and batch processing across our java projects.  Currently we are still in a phase of exploration. 

We have a few use cases where we have to read in big excel files or read out rather large source tables to move data to other files or database(s).  Since we like to move away from Sprint Batch and Quartz, there is a need to be able to use chunking.  Sending 10000's of insert statements one by one to the database would be a no-go.  You usually read in a chunk of for example 1000 records in memory, transform them and then write them to a file or db in one insert statement of 1000 records.

Sprint batch offers this currently out of the box but with a few other major setbacks being very obese in dependencies and complexity.  So we like to move to Jobrunr, however is this possible at the moment ?  Is there a way to save in a running job what rownum is next or how many chunks already have passed, suppose we want to retry a failed job ?  Is there some way to save a job context ?  If it's in already could you point me to the right manual section I'm overlooking ?

I would really appreciate any help provided, even an answer akin to you have to develop it yourself would be appreciated.

Kind regards

**Comment by rdehuyss** [comment, marked as answer]
Hi @Celludriel ,

we've definitely seen this use case before and there are some ways to solve it but it often comes down to a DIY solution. 

Personally, for ETL types of jobs, I've used the following approach which worked out great (using JobRunr Pro though). I start a Batch Job which
  - starts a first job that determines the amount of rows to be altered during the ETL job (let's say for the sake of the example 10_000_000)
  - then start a Batch Job where  launch 10 child jobs that each processes 1_000_000 jobs (e.g. first child job from 0 to 1_000_000, second child job from 1_000_000 to 2_000_000, ...) where the last one will do more if any new rows have been added. The amount of child jobs (in this example 10) depends on the load and memory pressure your DB and `BackgroundJobServer` can handle. For the ETL chunking and transforming, I can highly recommend Spring Data JDBC. 
  - if necessary, you can run a summary job when all child jobs complete, using `continueWith` from JobRunr Pro.
  - the nice thing is that as soon as all the child jobs finish, the parent batch succeeds and this can easily be monitored in the dashboard.

You could also do all of this in one single job but you loose the advantage of splitting up the larger job in different child jobs.

**Reply by Celludriel** [reply]
Thank you for this explanation, I just heard one of my co workers, was in your Devoxx talk earlier this month and I have a meeting with infrastructure this afternoon trying to defend the usage of Jobrunr @ De Lijn.  Can't promise anything yet but if talks go well we might get in contact with you.

Kind regards

**Reply by rdehuyss** [reply]
If there is anything I can do to help, do not hesitate to reach out!
